### PR TITLE
Feat: 쇼핑몰 리뷰 평점 예측 모델 구현

### PR DESCRIPTION
**Pull request by YujinJI**
### MAX_SEQUENCE_LENGTH
- `MAX_SEQUENCE_LENGTH`를 10으로 지정했습니다.
  - 학습 데이터와 테스트 데이터 단어 길이를 출력해 보았을 때 mean 값이 각각 8.6, 8.5로 나타남

### Label
- 평점 label은 1, 2, 4, 5로 총 4개입니다.
- `LabelEncoder`를 통해 [0, 1, 2, 3]의 값으로 수정한 뒤, keras의 `utils.to_categorical()`를 통해 0~1사이의 값으로 변경시켰습니다.

### EarlyStopping, ModelCheckpoint
- 학습을 진행할수록 점수가 떨어지는 것을 확인한 뒤 `EarlyStopping`과 `ModelCheckpoint`를 통해 베스트 모델 저장, 조기 종료될 수 있도록 했습니다.

### y_pred
- 0에서 1사이의 실수 값이 예측 값으로 나타남에 따라 `np.argmax()`를 통해 1, 2, 4, 5로 변경시킨 뒤 submission.csv 에서 'target' 열에 값을 넣었습니다.

### 주저리주저리
- 토큰화를 학습 데이터 기준으로 했었어서 테스트 데이터에 대해서 따로 토큰화를 진행시킨 후 데이콘에 제출하니 점수가 떨어졌습니다. *(0.01차잌)*
- 이전 submission.csv 파일과 겹치는 개수를 세어보니 25000개 중 20853으로 약 5000개 차이나길래 결과가 많이 다를 줄 알았쬽
  image

### CODE REVIEW

**Comment by BluePinetree** [comment]
_Commented on notebook `week7/지유진/쇼핑몰리뷰평점예측.ipynb` Cell 77 Line 5_
```python
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
train_y_encoded = encoder.fit_transform(train_y)
test_y_encoded = encoder.fit_transform(test_y)
```
1,2,4,5를 그대로 사용하면 원 핫 인코딩 시에 혼동이 생겨 나중에 예측할 때에 오류 혹은 추가적인 작업을 필요로 합니다. `LabelEncoder`를 적절히 잘 사용하였습니다.

다만, `scikit-learn`에서 Label Encoding과 같은 전처리를 진행할 시에 주의할 점이 있습니다.
`fit()`메서드는 반드시 훈련 데이터셋에만 적용시켜야 하기때문에 다음 코드와 같이 훈련 데이터셋에 적용시킨 전처리기를 바탕으로 시험 데이터셋에는 `transform()`만 호출하는 것이 적절할 듯 합니다.
```python
encoder = LabelEncoder()
train_y_encoded = encoder.fit_transform(train_y)
test_y_encoded = encoder.transform(test_y)
```

_Respond and view the context [here](https://gitnotebooks.com/SPAI-Lab/GenerativeAI_Study/pull/31#comment-1165)._

**Comment by BluePinetree** [comment]
### MAX_SEQUENCE_LENGTH
* 띄어쓰기 기준 단어 수의 평균이 시험/훈련 데이터 모두 8.5대가 나오니 어느 정도 여유를 주고 `MAX_SEQUENCE_LENGTH`를 10으로 설정한 것을 확인했습니다. Good👍

###  Label
* 설명에는 클래스가 총 1,2,3,4,5로 5개라고 하였지만 데이터 분석을 진행해보니 1,2,4,5로 4개만 존재하고 있습니다.
* 전처리하신대로 1,2,4,5를 그대로 사용하면 원핫 인코딩 시에 혼동이 생겨 나중에 예측할 때에 오류 혹은 추가적인 작업을 필요로 합니다. `LabelEncoder`를 적절히 잘 사용하였습니다.
* 비교적 간단한 작업이기 때문에 `LabelEncoder`를 사용하지 않고 다음과 같이 바로 mapping을 시켜도 문제가 없을 것 같습니다.
```python
train_y.map({0:1, 1:2, 2:4, 3:5})
``` 

### y_pred
* 확인하였습니다. 시행한 전처리에 적합한 후처리를 잘 진행해 주었습니다.

### 주저리주저리
* 성능을 올리기 위해 가설을 세운 뒤 이에 대해 분석하여 그 결과를 확인해보았군여. 아주 좋슴다. 👍
* 위의 코드에서 언급한 것과 마찬가지로,  tokenizer와 같은 전처리기는 훈련데이터가 아닌 시험 데이터에 개별적으로 처리할 시에 데이터 조작에 해당되기 때문에 혹시나 논문 작성 시나 대회 참여시에 주의가 필요합니다.
* 전처리기를 적합(fit)시키는 과정은 오직 훈련데이터에만 사용할 수 있고, 시험 데이터에는 최종 테스트용으로 이미 적합시킨 전처리기를 활용만 할 수 있다는 것을 기억해주시면 되겠습니다.
